File: 2022/src/bin/12.rs
use std::collections::VecDeque;

#[aoc2022::main(12)]
fn main(input: &str) -> (usize, usize) {
    let rows = input
        .split('\n')
        .map(|row| row.chars().collect::<Vec<char>>())
        .collect::<Vec<Vec<char>>>();

    let width = rows[0].len();
    let height = rows.len();

    let adj = calc_adj(height, width);

    let mut graph = rows.iter().flatten().copied().collect::<Vec<_>>();

    let start = graph.iter().position(|&cell| cell == 'S').unwrap();
    let end = graph.iter().position(|&cell| cell == 'E').unwrap();

    graph[start] = 'a';
    graph[end] = 'z';

    let size = width * height;

    let part_one = bfs(start, end, &adj, &graph, size);

    let part_two = bfs_backwards(end, &adj, &graph, size).unwrap();

    (part_one, part_two)
}

type Adj = Vec<Vec<usize>>;

fn normal(x: usize, y: usize, width: usize) -> usize {
    x + y * width
}

fn calc_adj(height: usize, width: usize) -> Adj {
    let mut adj = vec![vec![]; height * width];

    for y in 0..height {
        for x in 0..width {
            let index = normal(x, y, width);
            if x > 0 {
                adj[index].push(normal(x - 1, y, width));
            }
            if y + 1 < height {
                adj[index].push(normal(x, y + 1, width));
            }
            if x + 1 < width {
                adj[index].push(normal(x + 1, y, width));
            }
            if y > 0 {
                adj[index].push(normal(x, y - 1, width));
            }
        }
    }

    adj
}

fn bfs(root: usize, end: usize, adj: &Adj, graph: &[char], size: usize) -> usize {
    let mut distances = vec![0; size];

    let mut visited = vec![false; size];

    let mut q = VecDeque::new();

    visited[root] = true;

    q.push_back(root);

    distances[root] = 0;
    distances[end] = size;

    loop {
        let current = q.pop_front();

        match current {
            None => {
                break;
            }
            Some(elem) => {
                for &vec in adj[elem].iter() {
                    if visited[vec] {
                        continue;
                    }

                    if (graph[vec] as usize) <= (graph[elem] as usize) + 1 {
                        distances[vec] = distances[elem] + 1;
                        visited[vec] = true;

                        if vec != end {
                            q.push_back(vec);
                        }
                    }
                }
            }
        }
    }

    distances[end]
}

fn bfs_backwards(end: usize, adj: &Adj, graph: &[char], size: usize) -> Option<usize> {
    let mut distances = vec![0; size];

    let mut visited = vec![false; size];

    let mut q = VecDeque::new();

    visited[end] = true;

    q.push_back(end);

    let mut shortest: Option<usize> = None;

    loop {
        let current = q.pop_front();

        match current {
            None => {
                break;
            }
            Some(elem) => {
                for &vec in adj[elem].iter() {
                    if visited[vec] {
                        continue;
                    }

                    if (graph[vec] as usize) + 1 >= (graph[elem] as usize) {
                        distances[vec] = distances[elem] + 1;

                        visited[vec] = true;

                        if graph[vec] != 'a' {
                            q.push_back(vec);
                        } else {
                            match shortest {
                                None => shortest = Some(distances[vec]),
                                Some(n) if distances[vec] < n => {
                                    shortest = Some(distances[vec]);
                                }
                                _ => {}
                            }
                        }
                    }
                }
            }
        }
    }

    shortest
}

File: 2022/swift/12.swift
import Foundation

let filename = "./input/12.in"

func coordToIndex(_ coords: (x: Int, y: Int), _ width: Int) -> Int {
    coords.y * width + coords.x
}

func indexToCoord(_ index: Int, _ width: Int) -> (x: Int, y: Int) {
    (index % width, index / width)
}

// create adjacent matrix
func calcAdj<T>(_ graph: [T], _ width: Int) -> [[Int]] {
    var adjacent: [[Int]] = []

    let height = graph.count / width

    for (index, _) in graph.enumerated() {
        var localAdj: [Int] = []
        let (x, y) = indexToCoord(index, width)

        let up = (x, y - 1)
        let down = (x, y + 1)
        let left = (x - 1, y)
        let right = (x + 1, y)

        if up.1 >= 0 {
            localAdj.append(coordToIndex(up, width))
        }

        if down.1 < height {
            localAdj.append(coordToIndex(down, width))
        }

        if left.0 >= 0 {
            localAdj.append(coordToIndex(left, width))
        }

        if right.0 < width {
            localAdj.append(coordToIndex(right, width))
        }

        adjacent.append(localAdj)
    }

    return adjacent
}

struct Queue<Element> {
    var items: [Element] = []
    mutating func push(_ item: Element) {
        items.append(item)
    }

    mutating func pop() -> Element {
        return items.removeLast()
    }

    mutating func front() -> Element {
        return items.remove(at: 0)
    }

    func empty() -> Bool {
        return items.count == 0
    }
}

func bfs(_ root: Int, _ end: Int, _ graph: [Character], _ adj: [[Int]]) -> Int {
    var queue = Queue<Int>()
    var visited = graph.map { _ in
        false
    }

    var distances: [Int] = graph.map { _ in 0 }

    visited[root] = true
    distances[root] = 0
    distances[end] = graph.count

    queue.push(root)

    while !queue.empty() {
        let next = queue.front()

        for node in adj[next] {
            if visited[node] { continue }

            let peak = graph[node].asciiValue!
            let current = graph[next].asciiValue!

            if peak <= current + 1 {
                visited[node] = true

                distances[node] = distances[next] + 1

                if node != end {
                    queue.push(node)
                }
            }
        }
    }

    return distances[end]
}

func bfs_backwards(_ end: Int, _ graph: [Character], _ adj: [[Int]]) -> Int? {
    var queue = Queue<Int>()
    var visited = graph.map { _ in false }

    var distances = graph.map { _ in 0 }

    var shortest: Int?

    visited[end] = true
    distances[end] = 0

    queue.push(end)

    while !queue.empty() {
        let next = queue.front()

        for node in adj[next] {
            if visited[node] { continue }

            let peak = graph[node].asciiValue!
            let current = graph[next].asciiValue!

            if peak + 1 >= current {
                visited[node] = true

                distances[node] = distances[next] + 1

                if graph[node] != "a" {
                    queue.push(node)
                } else {
                    if shortest == nil || distances[node] < shortest! {
                        shortest = distances[node]
                    }
                }
            }
        }
    }

    return shortest
}

func flatGraph<T>(_ graph: [[T]]) -> [T] {
    var flat: [T] = []

    for row in graph {
        flat.append(contentsOf: row)
    }

    return flat
}

func main() {
    do {
        let rows: [[Character]] = try String(contentsOfFile: filename).split(separator: "\n").map { Array(String($0)) }

        let width = rows[0].count

        var root = 0
        var end = 0

        rows.enumerated().forEach { entry in
            let (y, row) = entry
            row.enumerated().forEach { cell in
                let (x, value) = cell

                if value == "S" {
                    let index = coordToIndex((x, y), width)
                    root = index
                }

                if value == "E" {
                    end = coordToIndex((x, y), width)
                }
            }
        }

        var graph = flatGraph(rows)

        graph[root] = "a"
        graph[end] = "z"

        let adj = calcAdj(graph, width)

        let shortestFromRoot = bfs(root, end, graph, adj)

        print("Part one:", shortestFromRoot)

        let shortestFromEnd = bfs_backwards(end, graph, adj)!

        print("Part two:", shortestFromEnd)
    } catch {
        print(error)
    }
}

main()
